File: crates/kuben-core/src/perm.rs
//! Permissions and built-in roles.

use std::{fmt, str::FromStr};

use serde::{Deserialize, Serialize};

/// Fine-grained permission. Every mutating or streaming API requires one.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum Perm {
    OrgRead,
    OrgAdmin,
    ProjectRead,
    ProjectWrite,
    EnvRead,
    EnvWrite,
    EnvDeleteProtected,
    AppRead,
    AppWrite,
    AppDeploy,
    AppLogsRead,
    AppExec,
    SecretRead,
    SecretWrite,
    ReleasePromote,
    ReleaseApprove,
    AuditRead,
    UserAdmin,
}

/// Built-in roles. Custom roles map to a set of [`Perm`]s (future work).
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Role {
    Owner,
    Admin,
    Developer,
    Viewer,
}

impl Role {
    /// Permissions granted by this role.
    #[must_use]
    pub fn perms(self) -> &'static [Perm] {
        use Perm::{
            AppDeploy, AppExec, AppLogsRead, AppRead, AppWrite, AuditRead, EnvDeleteProtected, EnvRead,
            EnvWrite, OrgAdmin, OrgRead, ProjectRead, ProjectWrite, ReleaseApprove, ReleasePromote,
            SecretRead, SecretWrite, UserAdmin,
        };
        match self {
            Self::Owner => &[
                OrgRead,
                OrgAdmin,
                ProjectRead,
                ProjectWrite,
                EnvRead,
                EnvWrite,
                EnvDeleteProtected,
                AppRead,
                AppWrite,
                AppDeploy,
                AppLogsRead,
                AppExec,
                SecretRead,
                SecretWrite,
                ReleasePromote,
                ReleaseApprove,
                AuditRead,
                UserAdmin,
            ],
            Self::Admin => &[
                OrgRead,
                ProjectRead,
                ProjectWrite,
                EnvRead,
                EnvWrite,
                AppRead,
                AppWrite,
                AppDeploy,
                AppLogsRead,
                AppExec,
                SecretRead,
                SecretWrite,
                ReleasePromote,
                ReleaseApprove,
                AuditRead,
                UserAdmin,
            ],
            Self::Developer => &[
                OrgRead,
                ProjectRead,
                EnvRead,
                AppRead,
                AppWrite,
                AppDeploy,
                AppLogsRead,
                AppExec,
                SecretRead,
            ],
            Self::Viewer => &[OrgRead, ProjectRead, EnvRead, AppRead, AppLogsRead],
        }
    }

    /// Position in the strict hierarchy `viewer < developer < admin < owner`;
    /// every role's permissions are a superset of the role below it.
    #[must_use]
    pub const fn rank(self) -> u8 {
        match self {
            Self::Viewer => 0,
            Self::Developer => 1,
            Self::Admin => 2,
            Self::Owner => 3,
        }
    }

    /// The weaker of two roles (e.g. the effective role of an API token is
    /// the weaker of its own cap and its owner's role).
    #[must_use]
    pub const fn weaker(self, other: Self) -> Self {
        if self.rank() <= other.rank() { self } else { other }
    }

    /// Whether this role grants `perm`.
    #[must_use]
    pub fn grants(self, perm: Perm) -> bool {
        self.perms().contains(&perm)
    }
}

impl fmt::Display for Role {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            Self::Owner => "owner",
            Self::Admin => "admin",
            Self::Developer => "developer",
            Self::Viewer => "viewer",
        };
        f.write_str(s)
    }
}

impl FromStr for Role {
    type Err = crate::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "owner" => Ok(Self::Owner),
            "admin" => Ok(Self::Admin),
            "developer" => Ok(Self::Developer),
            "viewer" => Ok(Self::Viewer),
            other => Err(crate::Error::Validation(format!("unknown role `{other}`"))),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn owner_is_superset_of_every_role() {
        for role in [Role::Admin, Role::Developer, Role::Viewer] {
            for perm in role.perms() {
                assert!(
                    Role::Owner.grants(*perm),
                    "owner must grant {perm:?} (from {role})"
                );
            }
        }
    }

    #[test]
    fn roles_form_a_strict_hierarchy() {
        let ordered = [Role::Viewer, Role::Developer, Role::Admin, Role::Owner];
        for pair in ordered.windows(2) {
            let (lower, higher) = (pair[0], pair[1]);
            assert!(lower.rank() < higher.rank());
            for perm in lower.perms() {
                assert!(
                    higher.grants(*perm),
                    "{higher} must grant {perm:?} (from {lower})"
                );
            }
        }
        assert_eq!(Role::Owner.weaker(Role::Developer), Role::Developer);
        assert_eq!(Role::Viewer.weaker(Role::Admin), Role::Viewer);
    }

    #[test]
    fn viewer_cannot_exec_or_write() {
        assert!(!Role::Viewer.grants(Perm::AppExec));
        assert!(!Role::Viewer.grants(Perm::AppWrite));
        assert!(Role::Viewer.grants(Perm::AppLogsRead));
    }

    #[test]
    fn role_parses_from_str() {
        assert_eq!("developer".parse::<Role>().expect("parse"), Role::Developer);
        assert!("root".parse::<Role>().is_err());
    }
}

File: apps/kuben/internal/core/perm/perm.go
// Package perm has the permissions and the built-in roles.
package perm

import (
	"slices"

	"github.com/Teamtem-dev/kuben/apps/kuben/internal/core/kerrors"
)

// Perm is a fine-grained permission. Every mutating or streaming API
// requires one.
type Perm string

// The permissions.
const (
	OrgRead            Perm = "org-read"
	OrgAdmin           Perm = "org-admin"
	ProjectRead        Perm = "project-read"
	ProjectWrite       Perm = "project-write"
	EnvRead            Perm = "env-read"
	EnvWrite           Perm = "env-write"
	EnvDeleteProtected Perm = "env-delete-protected"
	EnvProtect         Perm = "env-protect" // Weaken an environment's protection policy (M4.1).
	AppRead            Perm = "app-read"
	AppWrite           Perm = "app-write"
	AppDeploy          Perm = "app-deploy"
	AppLogsRead        Perm = "app-logs-read"
	AppExec            Perm = "app-exec"
	SecretRead         Perm = "secret-read"
	SecretWrite        Perm = "secret-write"
	ReleasePromote     Perm = "release-promote"
	ReleaseApprove     Perm = "release-approve"
	AuditRead          Perm = "audit-read"
	UserAdmin          Perm = "user-admin"
)

// Role is a built-in role. Custom roles map to a set of [Perm]s (future work).
type Role string

// The roles, in the strict hierarchy viewer < developer < admin < owner:
// every role's permissions are a superset of the role below it.
const (
	Viewer    Role = "viewer"
	Developer Role = "developer"
	Admin     Role = "admin"
	Owner     Role = "owner"
)

var viewerPerms = []Perm{OrgRead, ProjectRead, EnvRead, AppRead, AppLogsRead}

var developerPerms = append(slices.Clone(viewerPerms), AppWrite, AppDeploy, AppExec, SecretRead)

var adminPerms = append(slices.Clone(developerPerms),
	ProjectWrite, EnvWrite, SecretWrite, ReleasePromote, ReleaseApprove, AuditRead, UserAdmin)

var ownerPerms = append(slices.Clone(adminPerms), OrgAdmin, EnvDeleteProtected, EnvProtect)

// ParseRole reads a role name.
func ParseRole(s string) (Role, error) {
	switch r := Role(s); r {
	case Viewer, Developer, Admin, Owner:
		return r, nil
	}
	return "", kerrors.New(kerrors.Validation, "unknown role `%s`", s)
}

// Perms are the permissions the role grants; nil for an unknown role.
func (r Role) Perms() []Perm {
	switch r {
	case Viewer:
		return viewerPerms
	case Developer:
		return developerPerms
	case Admin:
		return adminPerms
	case Owner:
		return ownerPerms
	}
	return nil
}

// Rank is the role's position in the hierarchy; -1 for an unknown role, so
// that it is weaker than every role and grants nothing.
func (r Role) Rank() int {
	switch r {
	case Viewer:
		return 0
	case Developer:
		return 1
	case Admin:
		return 2
	case Owner:
		return 3
	}
	return -1
}

// Weaker is the weaker of two roles: the effective role of an API token is
// the weaker of its own cap and its owner's role.
func (r Role) Weaker(other Role) Role {
	if r.Rank() <= other.Rank() {
		return r
	}
	return other
}

// Grants reports whether the role grants p.
func (r Role) Grants(p Perm) bool { return slices.Contains(r.Perms(), p) }

func (r Role) String() string { return string(r) }

// UnmarshalText accepts only the four role names, as serde did: JSON, TOML
// and configuration decoding all go through it.
func (r *Role) UnmarshalText(text []byte) error {
	parsed, err := ParseRole(string(text))
	if err != nil {
		return err
	}
	*r = parsed
	return nil
}
